### DISCUSSION TITLE
"use client" not working as expected

### DISCUSSION THREAD

**Discussion by arthureberledev** [vercel/next.js#44663] [Help]
Hello,

i have recently found a new way to declare components that I like because of the improved auto completion. The problem with this is, that compared to the traditional way, "use client" is not working as expected:

Lets say I have a dummy carousel component that acts as a wrapper:

```
"use client";

type CarouselProps = HTMLAttributes<HTMLDivElement> & {};

export const Carousel = ({ children, ...props }: HTMLAttributes<HTMLDivElement>) => {
  const ref = useRef<HTMLDivElement>(null);

  return (
    <div className="carousel-slider" ref={ref} {...props}>
      {children}
    </div>
  );
};
[...]
```

**Example 1:**

If I now want to add a pagination component with the new approach that i have found, i get the error "Error: Unsupported Server Component type: undefined":

```
[...]
type CarouselPaginationProps = HTMLAttributes<HTMLDivElement> & {};

Carousel.Pagination = ({ children, className, ...props }: CarouselPaginationProps) => {
  return (
    <div className="carousel-pagination" {...props}>
      {children}
    </div>
  );
};
```

**Example 2:**

With the traditional approach, everything works fine and i get no error:

```
[...]
export const CarouselPagination = ({ children, className, ...props }: CarouselPaginationProps) => {
  return (
    <div className="carousel-pagination" {...props}>
      {children}
    </div>
  );
};
```

Repo: https://github.com/arthureberle/client-components-error

Is there any way to declare new components like in Example 1 without marking the parent component with "use client" as well?

What is the reason for this problem?

Big thanks in advance

**Comment by GabenGar** [comment]
I don't see the type info for `Carousel` signature provided in your snippets, so I assume there is none. Therefore typescript throws because you are trying to assign a member on the symbol which wasn't declared beforehand.

**Reply by arthureberledev** [reply]
Thank you for your Answer! But unfortunately I am having some problems understanding it I think. I have tried to properly type the `Carousel` Component:

```
type Carousel = {
  ({ children, ...props }: HTMLAttributes<HTMLDivElement>): JSX.Element;
  Pagination({ children, className, ...props }: HTMLAttributes<HTMLDivElement>): JSX.Element;
};
```

But sadly it is still not working. I have now also created a small GitHub Project: https://github.com/arthureberle/client-components-error

**Reply by GabenGar** [reply]
`Pagination` still isn't an attribute and that's not how you describe functions in `type`s. Not to mention you have to use `interface`s for these types of mixed signature functions.
Something like this:
```typescript
interface ICarousel {
  (props: HTMLAttributes<HTMLDivElement>): JSX.Element;
  Pagination: (props: HTMLAttributes<HTMLDivElement>) => JSX.Element
}
```
But really don't do these weird "functions with attributes" constructs, especially on react components. It's not that convenient versus extra import statements, but forces you to create full function signatures instead of just `props` argument. And also makes your code unshakeable.
